### PR TITLE
Implemented logit processors and stop criteria's

### PR DESCRIPTION
**Pull request by Maximilian-Winter**
I made a implementation of logit processors and custom stop criteria processors.

### CODE REVIEW

**Comment by Maximilian-Winter** [comment]
They are needed for proper guidance support:
https://github.com/microsoft/guidance/pull/70

File: llama_cpp/llama.py [hunk 1]
```diff
@@ -316,9 +316,14 @@ def _sample(
         mirostat_tau: llama_cpp.c_float,
         mirostat_eta: llama_cpp.c_float,
         penalize_nl: bool = True,
+        logits_processors=None
```
Comment (abetlen) [inline llama_cpp/llama.py:319]: Add type signature

File: llama_cpp/llama.py [hunk 2]
```diff
@@ -328,6 +333,10 @@ def _sample(
             else last_n_tokens_size
         )
         logits = self.eval_logits[-1]
+        for processor in logits_processors:
+            logits = processor(list(self.eval_tokens), logits)
+
+        self.eval_logits[-1] = logits
```
Comment (abetlen) [inline llama_cpp/llama.py:339]: Does logits_processor have to modify the logits, can it simply update the logits array here during the sampling?

File: llama_cpp/llama.py [hunk 3]
```diff
@@ -436,6 +445,8 @@ def sample(
         mirostat_eta: float = 0.1,
         mirostat_tau: float = 5.0,
         penalize_nl: bool = True,
+        logits_processors=None
```
Comment (abetlen) [inline llama_cpp/llama.py:448]: Add type signature

File: llama_cpp/llama.py [hunk 4]
```diff
@@ -484,6 +497,7 @@ def generate(
         mirostat_mode: int = 0,
         mirostat_tau: float = 5.0,
         mirostat_eta: float = 0.1,
+        logits_processors=None
```
Comment (abetlen) [inline llama_cpp/llama.py:500]: Add type signature

Comment (abetlen) [inline llama_cpp/llama.py:500]: Can we move `stopping_criteria` here so it matches the hugginface `generate` API? Then only use `generate` in the guidance PR.

**Review by abetlen — CHANGES_REQUESTED** [review]

File: llama_cpp/llama.py (hunk 2 above)
Comment (Maximilian-Winter) [reply llama_cpp/llama.py:338]: Yes the logits does have to modify the logits.

File: llama_cpp/llama.py (hunk 3 above)
Comment (Maximilian-Winter) [reply llama_cpp/llama.py]: I added a type signature.

File: llama_cpp/llama.py (hunk 4 above)
Comment (Maximilian-Winter) [reply llama_cpp/llama.py]: I added a type signature.

Comment (Maximilian-Winter) [reply llama_cpp/llama.py]: I would have to move the stop criteria check down here. Right now I'm using create completion in guidance.

File: llama_cpp/llama.py [hunk 5]
```diff
@@ -445,7 +444,7 @@ def sample(
         mirostat_eta: float = 0.1,
         mirostat_tau: float = 5.0,
         penalize_nl: bool = True,
-        logits_processors=None
+        logits_processors:  List[Callable[[List[llama_cpp.c_int], List[llama_cpp.c_float]], List[float]]] = None
```
Comment (abetlen) [inline llama_cpp/llama.py]: Name should be changed to `logits_processor` to match hugginface generate argument in https://huggingface.co/docs/transformers/v4.26.0/en/main_classes/text_generation#transformers.GenerationMixin.generate.logits_processor

Additionally, the types here should be changed to python `float` and `int`

**Review by abetlen — CHANGES_REQUESTED** [review]
